46.md

```diff
@@ -0,0 +1,159 @@
+`draft` `optional` `author:tiero` `author:giowe` `author:vforvalerio87`
+
+## Rationale
+
+Private keys should be exposed to as few systems - apps, operating systems, devices - as possible as each system adds to the attack surface.
+
+Entering private keys can also be annoying and requires exposing them to even more systems such as the operating system's clipboard that might be monitored by malicious apps.
+
+
+## Terms
+
+* **App**: Nostr app on any platform that *requires* to act on behalf of a nostr account.
+* **Signer**: Nostr app that holds the private key of a nostr account and *can sign* on its behalf.
+
+
+## `TL;DR`
+
+
+**App** and **Signer** sends ephemeral encrypted messages to each other using kind `24133`, using a relay of choice. 
+
+App prompts the Signer to do things such as fetching the public key or signing events.
+
+The `content` field must be an encrypted JSONRPC-ish **request** or **response**.
+
+## Signer Protocol
+
+### Messages
+
+#### Request
+
+```json
+{
+  "id": <random_string>,
+  "method": <one_of_the_methods>,
+  "params": [<anything>, <else>]
+}
+```
+
+#### Response
+
+```json
+{
+  "id": <request_id>,
+  "result": <anything>,
+  "error": <reason>
+}
+```
+
+### Methods
+
+
+#### Mandatory 
+
+These are mandatory methods the remote signer app MUST implement:
+
+- **get_public_key**
+  - params []
+  - result `pubkey` 
+- **sign_event**
+  - params [`event`]
+  - result `signature` 
+
+#### optional
+
+
+- **connect**
+  - params [`pubkey`]
+- **disconnect**
+  - params []
+- **delegate** 
+  - params [`pubkey`, `conditions query string`]
+  - result `nip26 delegation token`
+- **get_relays**
+  - params []
+  - result `{ [url: string]: {read: boolean, write: boolean} }` 
+- **nip04_encrypt**
+  - params [`pubkey`, `plaintext`]
+  - result `nip4 ciphertext`
+- **nip04_decrypt**
+  - params [`pubkey`, `nip4 ciphertext`]
+  - result [`plaintext`]
+
+
+NOTICE: `pubkey` and `signature` are hex-encoded strings.
+
+
+### Nostr Connect URI
+
+**Signer** discovers **App** by scanning a QR code, clicking on a deep link or copy-pasting an URI.
+
+The **App** generates a special URI with prefix `nostrconnect://` and base path the hex-encoded `pubkey` with the following querystring parameters **URL encoded**
+
+- `relay` URL of the relay of choice where the **App** is connected and the **Signer** must send and listen for messages.
+- `metadata`  metadata JSON of the **App** 
+    - `name` human-readable name of the **App** 
+    - `url` (optional) URL of the website requesting the connection
+    - `description` (optional) description of the **App**
+    - `icons` (optional) array of URLs for icons of the **App**.
+
+#### JavaScript
+
+```js
+const uri = `nostrconnect://<pubkey>?relay=${encodeURIComponent("wss://relay.damus.io")}&metadata=${encodeURIComponent(JSON.stringify({"name": "Example"}))}`
+```
+
+#### Example
+```sh
+nostrconnect://foobar?relay=wss%3A%2F%2Frelay.damus.io&metadata=%7B%22name%22%3A%22Example%22%7D
```

Comment:
`foobar`? I thought it need to be the hex-encoded public key?

Reply:
```suggestion
nostrconnect://b889ff5b1513b641e2a139f661a661364979c5beee91842f8f0ef42ab558e9d4?relay=wss%3A%2F%2Frelay.damus.io&metadata=%7B%22name%22%3A%22Example%22%7D
```